# Laboratorio de regresión - 3

## Significancia de factores

|                |   |
:----------------|---|
| **Nombre**     Dayanni Godoy Rosales   |
| **Fecha**     25/08/2025 |   |
| **Expediente** |   749206|   |

Descarga el archivo de publicidad y carga los datos (Advertising.csv).

In [22]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score

In [24]:
data = pd.read_csv('Advertising.csv')
data

Unnamed: 0     TV  radio  newspaper  sales
0             1  230.1   37.8       69.2   22.1
1             2   44.5   39.3       45.1   10.4
2             3   17.2   45.9       69.3    9.3
3             4  151.5   41.3       58.5   18.5
4             5  180.8   10.8       58.4   12.9
..          ...    ...    ...        ...    ...
195         196   38.2    3.7       13.8    7.6
196         197   94.2    4.9        8.1    9.7
197         198  177.0    9.3        6.4   12.8
198         199  283.6   42.0       66.2   25.5
199         200  232.1    8.6        8.7   13.4

[200 rows x 5 columns]

**¿Hay alguna relación entre el presupuesto para publicidad y las ventas?**

Nuestra primera meta debe ser determinar si hay evidencia en los datos de que haya una asociación entre estas variables.

- ¿Por qué? ¿Qué resultaría si nos diéramos cuenta de la falta de relación entre el presupuesto de publicidad y las ventas?

Que no vale la pensa gastar en publicidad al no tener impacto en las ventas.

**¿Qué tan fuerte es esta relación?**
Asumiendo que existe esta relación, ¿nos sirve conocer el impacto que tiene invertir en publicidad en las ventas?

Si, para saber qué tanto vale la pena invertir.

**¿Cuáles medios están asociados con las ventas? ¿Qué tan grande es la asociación entre un medio específico y las ventas?**

Hay 3 medios distintos en los datos. ¿Sirve invertir en los 3? ¿Conviene más invertir sólo en uno?

**¿Qué tan seguros estamos de que podríamos predecir ventas futuras?**

**¿La relación es lineal?**

**¿Hay sinergia entre estos medios?**

Puede ser que gastar \\$50,000 en publicidad y otros \\$50,000 en radio es mejor opción que gastar \\$100,000 en televisión. A esto le llamamos efecto de interacción.

Podemos usar regresión lineal para responder todas estas preguntas.

Realiza una regresión lineal:

$$ \text{ventas} \approx \beta_0 + (\beta_1)(\text{TV})$$

In [68]:
x1 = data['TV'].values.reshape(-1,1)
y = data['sales']

In [72]:
x = x1
X = np.hstack([x])
lr = LinearRegression()
lr.fit(x,y)
b0 = lr.intercept_
print("Intercepto (b0):", b0)  # constante
b1 = lr.coef_
print("Coeficiente (b1):", b1)       # pendiente
R2= r2_score(y, lr.predict(x)) 
print("R2:",R2)   # error

Intercepto (b0): 7.032593549127693
Coeficiente (b1): [0.04753664]
R2: 0.611875050850071


### Verificando la precisión de nuestros coeficientes estimados

Recuerda que en el mundo real hay ruidos y errores de medición. Siempre se asume que la verdadera relación entre $X$ y $Y$ es $$Y = \beta_0 + \beta_1 X + \epsilon$$

Se asume que el término de error es independiente de $X$ (el error siempre es el mismo sin importar el valor de $X$). Este modelo describe a la *línea de regresión de la población*, que es la mejor aproximación de la verdadera relación entre $X$ y $Y$. Cuando usamos mínimos cuadrados encontramos la *línea de mínimos cuadrados*.

¿Cuál es la diferencia entre población y muestra?

La población es el total, la muestra es una parte de esa población total.

¿Cuál crees que sea la diferencia entre hacer una regresión con todos los datos de la población y una muestra de ella?

Regresión con población: Obtienes los parámetros verdaderos (β₀, β₁) de la relación exacta

Regresión con muestra: Obtienes estimaciones (β̂₀, β̂₁) que contienen error muestral

La línea de regresión de la población no se puede observar. El concepto de comparar estas líneas es una extensión natural del acercamiento estadístico estándar de usar información de una muestra para estimar características de una población grande.

Imagina que quieres encontrar la altura promedio de un mexicano $\mu$. Medir a todos y cada uno de los mexicanos en situaciones similares, con la misma regla, mismo operador, y otras incontables formas de minimizar la variación de la medida es una tarea imposible. Lo que podemos asumir es que $\hat{\mu} = \bar{y}$. La media poblacional y la media muestral son diferentes, pero la media muestral es usualmente un buen estimado.

De la misma manera, como no contamos con el 100% de la información para hacer una regresión, los coeficientes $\beta_0$ y $\beta_1$ son desconocidos. Podemos estimarlos usando mínimos cuadrados, encontrando $\hat{\beta_0}$ y $\hat{\beta_1}$. Puede que las muestras que tengamos en ese momento estén un poco por encima de la media, pero otras muestras en otro momento puede que estén debajo de la media. En general, esperamos que el promedio de las aproximaciones $\hat{\mu}$ aproxime a $\mu$.

Esto lleva a la pregunta: ¿qué tan cercanos son nuestros coeficientes estimados a los verdaderos coeficientes? Utilizamos el concepto de error estándar para evaluar esto.

$$ \text{Var}(\hat{\mu})=\text{SE}(\hat{\mu})^2 = \frac{\sigma^2}{n} $$

Donde $\sigma$ es la desviación estándar de cada una de las observaciones $y_i$ de $Y$. El error estándar nos dice la cantidad promedio que el estimado difiere del valor verdadero. Podemos ver en la fórmula que entre más observaciones tengamos el error se hace más pequeño. Las fórmulas para errores estándar de $\hat{\beta_0}$ y $\hat{\beta_1}$ son:

$$ \text{SE}(\hat{\beta_0})^2 = \sigma^2 [\frac{1}{n} + \frac{\bar{x}^2}{\sum_{i=1}^n (x_i - \bar{x})^2}]$$

$$ \text{SE}(\hat{\beta_1})^2 = \frac{\sigma^2}{\sum_{i=1}^n (x_i - \bar{x})^2}$$

$$ \sigma^2 = \text{Var}(\epsilon) = \text{RSE}^2 = \frac{\text{RSS}}{n-p}$$

Para que estas fórmulas sean validas asumimos que los errores $\epsilon_i$ tienen varianza común $\sigma^2$ y que no están correlacionados.

Calcula los errores estándar de los coeficientes

In [74]:
x_m = np.mean(x)
y_m = np.mean(y)
n = len(y)
p = 2
RSS = sum((lr.predict(x)-y)**2)
RSS

2102.5305831313512

In [76]:
sigma2 = RSS / (n - p)
sigma2

10.618841328946218

In [78]:
SEB0 = np.sqrt(sigma2 * ((1/n) + (x_m**2 /(sum((x - x_m)**2)))))
SEB0

array([0.45784294])

In [80]:
SEB1 = np.sqrt(sigma2 / (sum((x - x_m)**2)))
SEB1

array([0.00269061])

Estos errores se pueden usar para calcular intervalos de confianza. Un intervalo de confianza del $95\%$ se define como un rango de valores en el cuál se encuentra el desconocido valor verdadero con un $95\%$ de probabilidad.

Otra forma de verlo es que si tomamos muestras repetidas y construimos un intervalo de confianza para cada una, el $95\%$ de los intervalos creados van a contener el valor verdadero. Para la regresión el intervalo de confianza del $95\%$ toma la forma:

$$ \hat{\beta_j} \pm 2\text{SE}(\hat{\beta_j})$$

Calcula los intervalos de confianza para los coeficientes estimados:

In [84]:
b0_bajo = lr.intercept_-(2 * SEB0)
b0_alto = lr.intercept_+(2 * SEB0)
b1_bajo = lr.coef_-(2 * SEB1)
b1_alto = lr.coef_+(2 * SEB1)

In [86]:
(b0_bajo,b0_alto)

(array([6.11690767]), array([7.94827943]))

In [88]:
(b1_bajo,b1_alto)

(array([0.04215543]), array([0.05291785]))

Los errores estándar también se usan para realizar pruebas de hipótesis. La prueba de hipótesis más común es probar la hipótesis nula de:

$$ H_0: \text{No hay relación entre } X \text{ y } Y \ \ \ \ (\beta_1=0)$$

contra la hipótesis alternativa:
$$ H_i: \text{Hay alguna relación entre } X \text{ y } Y \ \ \ (\beta_1 \neq 0)$$

Explica con tus palabras el significado de la hipótesis nula y la hipótesis alternativa.

Hipotesis nula: B1 = 0,  la pendiente de la regresión es cero, entonces no hay efecto real de x sobre y.

Hipotesis alterna:  si B diferente a 0,entonces el efecto entre x y y es real.

Para probal la hipótesis nula debemos determinar si nuestro estimado $\hat{\beta_1}$ de $\beta_1$ está lo suficientemente alejado de cero para que podamos decir con confianza que este valor no es cero. 

¿Qué tan lejos? Depende de qué tanta confianza tengamos en el estimado encontrado. Si nuestro error estándar es pequeño y nuestro estimado está alejado de cero podríamos decir que hay muy poca probabilidad de que el valor verdadero sea 0. En cambio, si nuestro error estándar es grande y nuestro estimado está muy cerca de cero, entonces podrías ser que el valor verdadero sea cero y que no haya relación entre las variables.

Se calcula un *estadístico t* dado por
$$ t = \frac{\hat{\beta_j} - \mu}{\text{SE}(\hat{\beta_j})} $$

donde $\mu$ es el valor contra el que queremos probar.

Calcula el estadístico t para tus coeficientes estimados, usando como referencia la prueba de hipótesis.

In [98]:
t_b0 = lr.intercept_/SEB0
t_b1=lr.coef_/SEB1
t_b1,t_b0

(array([17.6676256]), array([15.36027517]))

La distribución t tiene forma de campana y se parece bastante a la distribución normal cuando $n > 30$. Ya sólo es cuestión de calcular la probabilidad de observar cualquier número tal que su valor absoluto sea igual o mayor que el valor absoluto del estadístico t calculado. En otras palabras:
$$ P(|x| \geq |t|) $$

A esta probabilidad la llamamos *p-value*. Un *p-value* pequeño indica que es poco probable que exista por puro azar una relación significativa entre predictor y respuesta, en caso de que no haya una asociación real entre predictor y respuesta. En otras palabras, el *p-value* te dice la probabilidad de que parezca que hay relación cuando no la hay.

Si el *p-value* es pequeño, inferimos que sí hay una asociación entre el predictor y la respuesta, y **rechazamos la hipótesis nula**.
  

¿Qué tan pequeño? Depende de la aplicación. Un valor muy común es del $5\%$.

Utiliza el siguiente código para calcular el *p-value* para tus coeficientes

`from scipy import stats`

`p_bj = 2*(1 - stats.t.cdf(np.abs(t_bj), n-p))`

In [103]:
from scipy import stats

In [105]:
p_b0 = 2*(1 - stats.t.cdf(np.abs(t_b0), n-p))
p_b0

array([0.])

In [107]:
p_b1 = 2*(1 - stats.t.cdf(np.abs(t_b1), n-p))
p_b1

array([0.])

¿Se rechaza la hipótesis nula? ¿Qué significa?

Si es diferente a 0, si es 0.047 entonces TV si afecta en sales.

Realiza otras dos regresiones. Ya tienes hecha la regresión de ventas dado el gasto en publicidad de TV. Realiza la regresión para gastos en radio y gastos en periódico. Organiza las respuestas para que debajo de esta celda se tenga:
- Título de regresión
- Coeficientes estimados
- Errores estándar de los coeficientes
- Intervalos de confianza
- Estadísticos t
- p-values
- Observaciones

## Regresión radio

In [113]:
#Regresión radio
x2 = data['radio'].values.reshape(-1,1)
x = x2
lr = LinearRegression()
lr.fit(x,y)
b0 = lr.intercept_
print("Coeficientes estimados")
print("Intercepto:", b0)  # constante
b1 = lr.coef_
print("Coeficiente:", b1)       # pendiente
R2 = r2_score(y, lr.predict(x)) 
print("R2:",R2)   # error

# Errores estándar de los coeficientes}
print("Errores estándar de los coeficientes")
x_m  = np.mean(x)
y_m = np.mean(y)
n = len(y)
p = 2
RSS = sum((lr.predict(x)-y)**2)
sigma2 = RSS / (n - p)
SEB0 = np.sqrt(sigma2 * ((1/n) + (x_m**2 /(sum((x - x_m)**2)))))
SEB1 = np.sqrt(sigma2 / (sum((x - x_m)**2)))
print("SEB0:", SEB0)
print("SEB1:" ,SEB1)

# Intervalos de confianza
b0_bajo = lr.intercept_-(2 * SEB0)
b0_alto = lr.intercept_+(2 * SEB0)
b1_bajo = lr.coef_-(2 * SEB1)
b1_alto = lr.coef_+(2 * SEB1)
print("Intervalos de confianza b0: ", b0_bajo,b0_alto)
print("Intervalos de confianza b1: ", b1_bajo,b1_alto)

# Estadísticos t
t_b0 = lr.intercept_/SEB0
t_b1=lr.coef_/SEB1
print("Estadísticos t:",t_b1,t_b0)

# p-values
p_b1 = 2*(1 - stats.t.cdf(np.abs(t_b0), n-p))
p_b2 = 2*(1 - stats.t.cdf(np.abs(t_b1), n-p))
print("p-values,",p_b1,p_b2)

Coeficientes estimados
Intercepto: 9.311638095158283
Coeficiente: [0.20249578]
R2: 0.33203245544529525
Errores estándar de los coeficientes
SEB0: [0.5629005]
SEB1: [0.02041131]
Intervalos de confianza b0:  [8.1858371] [10.43743909]
Intervalos de confianza b1:  [0.16167317] [0.2433184]
Estadísticos t: [9.92076547] [16.54224531]
p-values, [0.] [0.]


### Observaciones:

**R2 es bajo:**  x solo explica el 33.2% de la variabilidad de y.

**p-values = 0.0:** Extremadamente significativos. La probabilidad de observar estos coeficientes por pura casualidad  cero.

**Rechazamos la hipótesis nula (β₁ = 0).** Existe una relación estadísticamente significativa entre X e Y.

**Error estándar de β₁ (SEB1 = 0.0204):** La estimación de la pendiente es muy precisa. Esperamos que varíe solo unas 0.02 unidades entre muestras.

**Intervalo de Confianza para β₁: [0.162, 0.243]:** Con un 95% de confianza, el efecto de X sobre Y se encuentra en este rango. Como el intervalo no incluye el cero, confirma la significancia.

## Regresión newspaper

In [117]:
#Regresión newspaper
x3 = data['newspaper'].values.reshape(-1,1)
x = x3
lr = LinearRegression()
lr.fit(x,y)
b0 = lr.intercept_
print("Coeficientes estimados")
print("Intercepto:", b0)  # constante
b1 = lr.coef_
print("Coeficiente:", b1)       # pendiente
R2 = r2_score(y, lr.predict(x)) 
print("R2:",R2)   # error

# Errores estándar de los coeficientes}
print("Errores estándar de los coeficientes")
x_m  = np.mean(x)
y_m = np.mean(y)
n = len(y)
p = 2
RSS = sum((lr.predict(x)-y)**2)
sigma2 = RSS / (n - p)
SEB0 = np.sqrt(sigma2 * ((1/n) + (x_m**2 /(sum((x - x_m)**2)))))
SEB1 = np.sqrt(sigma2 / (sum((x - x_m)**2)))
print("SEB0:", SEB0)
print("SEB1:" ,SEB1)

# Intervalos de confianza
b0_bajo = lr.intercept_-(2 * SEB0)
b0_alto = lr.intercept_+(2 * SEB0)
b1_bajo = lr.coef_-(2 * SEB1)
b1_alto = lr.coef_+(2 * SEB1)
print("Intervalos de confianza b0: ", b0_bajo,b0_alto)
print("Intervalos de confianza b1: ", b1_bajo,b1_alto)

# Estadísticos t
t_b0 = lr.intercept_/SEB0
t_b1=lr.coef_/SEB1
print("Estadísticos t:",t_b1,t_b0)

# p-values
p_b1 = 2*(1 - stats.t.cdf(np.abs(t_b0), n-p))
p_b2 = 2*(1 - stats.t.cdf(np.abs(t_b1), n-p))
print("p-values,",p_b1,p_b2)

Coeficientes estimados
Intercepto: 12.35140706927816
Coeficiente: [0.0546931]
R2: 0.05212044544430516
Errores estándar de los coeficientes
SEB0: [0.62142019]
SEB1: [0.01657572]
Intervalos de confianza b0:  [11.10856669] [13.59424744]
Intervalos de confianza b1:  [0.02154165] [0.08784454]
Estadísticos t: [3.29959074] [19.87609562]
p-values, [0.] [0.0011482]


### Observaciones:

**R2 es muuy bajo:**  x solo explica el 5% de la variabilidad de y.
Considero que lo demás es irrelevante considerando el bajo R2

## Regresión lineal múltiple

En lugar de hacer una regresión para cada factor independiente, quizás se puede extender el modelo para que tenga varios factores dentro:

$$ Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_p X_p + \epsilon $$

Para nuestro ejemplo de publicidad:

$$ \text{sales} = \beta_0 + \beta_1 (\text{TV}) + \beta_2 (\text{radio}) + \beta_3 (\text{newspaper}) + \epsilon $$

Utiliza la librería `statsmodels` para realizar la regresión. Por defecto la librería statsmodels no toma en cuenta el intercepto ($\beta_0$), por lo que se tendrá que agregar una columna de unos de tamaño *n* a la matriz X.

`import statsmodels.api as sm`

`ols = sm.OLS(Y, X)`

`results = ols.fit()`

`results.summary()`

In [123]:
import statsmodels.api as sm

### Regresión múltiple 3 variables

In [126]:
ones = np.ones([n,1])
X = np.hstack([ones,x1,x2,x3])

In [128]:
ols = sm.OLS(y, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sat, 30 Aug 2025   Prob (F-statistic):           1.58e-96
Time:                        10:38:05   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
x1             0.0458      0.001     32.809      0.000       0.043       0.049
x2             0.1885      0.009     21.893      0.000       0.172       0.206
x3            -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Regresión múltiple con TV * radio

In [131]:
ones = np.ones([n,1])
X = np.hstack([ones,x1,x2,(x1*x2)])

In [133]:
ols = sm.OLS(y, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.968
Model:                            OLS   Adj. R-squared:                  0.967
Method:                 Least Squares   F-statistic:                     1963.
Date:                Sat, 30 Aug 2025   Prob (F-statistic):          6.68e-146
Time:                        10:38:05   Log-Likelihood:                -270.14
No. Observations:                 200   AIC:                             548.3
Df Residuals:                     196   BIC:                             561.5
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.7502      0.248     27.233      0.000       6.261       7.239
x1             0.0191      0.002     12.699      0.000       0.016       0.022
x2             0.0289      0.009      3.241      0.001       0.011       0.046
x3             0.0011   5.24e-05     20.727      0.000       0.001       0.001
==============================================================================
Omnibus:                      128.132   Durbin-Watson:                   2.224
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1183.719
Skew:                          -2.323   Prob(JB):                    9.09e-258
Kurtosis:                      13.975   Cond. No.                     1.80e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.8e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Regresión múltiple 6 variables

In [136]:
ones = np.ones([n,1])
X = np.hstack([ones,x1,x2,x3,(x1*x2),(x1*x3),(x2*x3)])
X

array([[1.000000e+00, 2.301000e+02, 3.780000e+01, ..., 8.697780e+03,
        1.592292e+04, 2.615760e+03],
       [1.000000e+00, 4.450000e+01, 3.930000e+01, ..., 1.748850e+03,
        2.006950e+03, 1.772430e+03],
       [1.000000e+00, 1.720000e+01, 4.590000e+01, ..., 7.894800e+02,
        1.191960e+03, 3.180870e+03],
       ...,
       [1.000000e+00, 1.770000e+02, 9.300000e+00, ..., 1.646100e+03,
        1.132800e+03, 5.952000e+01],
       [1.000000e+00, 2.836000e+02, 4.200000e+01, ..., 1.191120e+04,
        1.877432e+04, 2.780400e+03],
       [1.000000e+00, 2.321000e+02, 8.600000e+00, ..., 1.996060e+03,
        2.019270e+03, 7.482000e+01]])

In [138]:
ols = sm.OLS(y, X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.969
Model:                            OLS   Adj. R-squared:                  0.968
Method:                 Least Squares   F-statistic:                     993.3
Date:                Sat, 30 Aug 2025   Prob (F-statistic):          3.68e-142
Time:                        10:38:06   Log-Likelihood:                -267.49
No. Observations:                 200   AIC:                             549.0
Df Residuals:                     193   BIC:                             572.1
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.4602      0.318     20.342      0.000       5.834       7.087
x1             0.0203      0.002     12.633      0.000       0.017       0.024
x2             0.0229      0.011      2.009      0.046       0.000       0.045
x3             0.0170      0.010      1.691      0.092      -0.003       0.037
x4             0.0011   5.72e-05     19.930      0.000       0.001       0.001
x5         -7.971e-05   3.58e-05     -2.227      0.027      -0.000   -9.12e-06
x6            -0.0001      0.000     -0.464      0.643      -0.001       0.000
==============================================================================
Omnibus:                      111.385   Durbin-Watson:                   2.222
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              767.483
Skew:                          -2.046   Prob(JB):                    2.20e-167
Kurtosis:                      11.681   Cond. No.                     3.80e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.8e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

¿Qué diferencias puedes observar entre los *p-values* de una regresión múltiple y los encontrados en las regresiones simples? ¿Por qué crees que existen estas diferencias?

Primero salta a la vista que el R2 aumentó a 0.897 (frente al 33% y 5% de las simples). El modelo conjunto explica casi el 90% de la variabilidad de sales.

Luego, los p-values: 
Regresión simple: Todas las variables individuales eran altamente significativas (p ≈ 0.0).

Regresión múltiple: x1 y x2 siguen siendo altamente significativas (p=0.000) pero x3 se vuelve no significativa (p=0.860), ¿por qué pasa esto?

(En esta parte me ayudó chat)
Colinealidad: Las variables predictoras están correlacionadas entre sí. Cuando se incluyen juntas en el modelo, compiten por explicar la misma variabilidad en sales.

Control de confusores: La regresión múltiple aísla el efecto único de cada variable, controlando por las demás. El efecto de x3 que se observaba en la regresión simple era espurio (capturaba efectos de x1 o x2).

Redundancia: x3 no aporta información nueva útil una vez que x1 y x2 están en el modelo. Su efecto aparente en la regresión simple era en realidad explicado por otras variables.

De cualquier manera, lo mejor en estos casos es hacer caso a la regresión multiple, la simple queda obsoleta.

## Referencia

James, G., Witten, D., Hastie, T., Tibshirani, R.,, Taylor, J. (2023). An Introduction to Statistical Learning with Applications in Python. Cham: Springer. ISBN: 978-3-031-38746-3

## Addendum

Para calcular los *p-values* de los parámetros sin `statsmodels`:
1. Calcular RSS
2. Calcular RSE
3. `var_beta = np.linalg.inv(X.T @ X) * rse**2` (X con columna de intercepto)
4. `std_beta = np.sqrt(var_beta.diagonal())` El orden de los valores corresponde al orden de los factores en las columnas de la matriz $X$.
5. Calcular *estadístico t*
6. Calcular *p-value*